File: go/src/enjarify-go/jvm/registers.go
// Copyright 2015 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package jvm

import (
	"enjarify-go/jvm/ir"
	"enjarify-go/util"
	"sort"
)

// As usual, assume no iincs

// A set of registers that currently are copies of each other.
type copySet struct {
	root ir.RegKey
	set  map[ir.RegKey]bool
	q    []ir.RegKey // keep track of insertion order in case root is overwritten
}

func newCopySet(key ir.RegKey) *copySet {
	return &copySet{key, map[ir.RegKey]bool{key: true}, []ir.RegKey{key}}
}
func (self *copySet) add(key ir.RegKey) {
	util.Assert(len(self.set) > 0)
	self.set[key] = true
	self.q = append(self.q, key)
}
func (self *copySet) remove(key ir.RegKey) {
	delete(self.set, key)
	// Heuristic - use oldest element still in set as new root
	for len(self.q) > 0 && !self.set[self.root] {
		self.root = self.q[0]
		self.q = self.q[1:]
	}
}
func (self *copySet) copy() *copySet {
	new := newCopySet(self.root)
	new.set = make(map[ir.RegKey]bool)
	for k, v := range self.set {
		new.set[k] = v
	}
	new.q = make([]ir.RegKey, 0, len(self.q))
	copy(new.q, self.q)
	return new
}

// Map registers to CopySets
type copySetsMap struct {
	lookup map[ir.RegKey]*copySet
}

func newCopySetsMap() *copySetsMap { return &copySetsMap{make(map[ir.RegKey]*copySet)} }

func (self *copySetsMap) get(key ir.RegKey) *copySet {
	if v, ok := self.lookup[key]; !ok {
		new := newCopySet(key)
		self.lookup[key] = new
		return new
	} else {
		return v
	}
}

func (self *copySetsMap) clobber(key ir.RegKey) {
	self.get(key).remove(key)
	delete(self.lookup, key)
}

func (self *copySetsMap) move(dest, src ir.RegKey) bool {
	// return false if the corresponding instructions should be removed
	s_set := self.get(src)
	d_set := self.get(dest)

	if s_set == d_set {
		// src and dest are copies of same value, so we can remove
		return false
	}
	d_set.remove(dest)
	s_set.add(dest)
	self.lookup[dest] = s_set
	return true
}
func (self *copySetsMap) load(key ir.RegKey) ir.RegKey {
	return self.get(key).root
}
func (self *copySetsMap) copy() *copySetsMap {
	copies := make(map[*copySet]*copySet)
	new := newCopySetsMap()
	for k, v := range self.lookup {
		if copies[v] == nil {
			copies[v] = v.copy()
		}
		new.lookup[k] = copies[v]
	}
	return new
}

func CopyPropagation(irdata *IRWriter) {
	instrs := irdata.Instructions

	replace := make(map[ir.Instruction][]ir.Instruction)
	single_pred_infos := make(map[ir.Instruction]*copySetsMap)
	prev := ir.Instruction(nil)

	current := newCopySetsMap()
	for _, instr := range instrs {
		// reset all info when control flow is merged
		if irdata.jump_targets[instr] {
			// try to use info if this was a single predecessor forward jump
			if prev != nil && !prev.Fallsthrough() && irdata.target_pred_counts[instr] == 1 {
				current = single_pred_infos[instr]
				if current == nil {
					current = newCopySetsMap()
				}
			} else {
				current = newCopySetsMap()
			}
		} else if ins, ok := instr.(*ir.RegAccess); ok {
			key := ins.RegKey
			if ins.Store {
				// check if previous instr was a load
				if prev, ok := prev.(*ir.RegAccess); ok && !prev.Store {
					if !current.move(key, prev.RegKey) {
						replace[prev] = nil
						replace[instr] = nil
					}
				} else {
					current.clobber(key)
				}
			} else {
				root_key := current.load(key)
				if key != root_key {
					_, ok := replace[instr]
					util.Assert(!ok)
					// replace with load from root register instead
					replace[instr] = []ir.Instruction{ir.NewRegAccess(root_key.Reg, root_key.T, false)}
				}
			}
		} else {
			for _, target := range instr.Targets() {
				label := irdata.Labels[target]
				if irdata.target_pred_counts[label] == 1 {
					single_pred_infos[label] = current.copy()
				}
			}
		}

		prev = instr
	}

	irdata.ReplaceInstrs(replace)
}

func isRemoveable(instr ir.Instruction) bool {
	// can remove if load or const since we know there are no side effects
	// note - instr may be nil
	switch instr := instr.(type) {
	case *ir.RegAccess:
		return !instr.Store
	case *ir.PrimConstant, *ir.OtherConstant:
		return true
	default:
		return false
	}
}

func RemoveUnusedRegisters(irdata *IRWriter) {
	// Remove stores to registers that are not read from anywhere in the method
	instrs := irdata.Instructions

	used := make(map[ir.RegKey]bool)
	for _, instr := range instrs {
		if ins, ok := instr.(*ir.RegAccess); ok && !ins.Store {
			used[ins.RegKey] = true
		}
	}

	replace := make(map[ir.Instruction][]ir.Instruction)
	prev := ir.Instruction(nil)
	for _, instr := range instrs {
		if ins, ok := instr.(*ir.RegAccess); ok && !used[ins.RegKey] {
			util.Assert(ins.Store)
			// if prev instruction is load or const, just remove it and the store
			// otherwise, replace the store with a pop
			if isRemoveable(prev) {
				replace[prev] = nil
				replace[instr] = nil
			} else {
				if ins.T.Wide() {
					replace[instr] = []ir.Instruction{ir.NewOther(POP2)}
				} else {
					replace[instr] = []ir.Instruction{ir.NewOther(POP)}
				}
			}

		}
		prev = instr
	}
	irdata.ReplaceInstrs(replace)
}

// Allocate registers to JVM registers on a first come, first serve basis
// For simplicity, parameter registers are preserved as is
func SimpleAllocateRegisters(irdata *IRWriter) {
	instrs := irdata.Instructions
	regmap := map[ir.RegKey]int{}
	for i, v := range irdata.initial_args {
		regmap[v] = i
	}
	next := len(irdata.initial_args)

	for _, instr := range instrs {
		if instr, ok := instr.(*ir.RegAccess); ok {
			if _, ok := regmap[instr.RegKey]; !ok {
				regmap[instr.RegKey] = next
				next++
				if instr.T.Wide() {
					next++
				}
			}
			_, ok = regmap[instr.RegKey]
			util.Assert(ok)
			instr.CalcBytecode(regmap[instr.RegKey])
		}
	}
	irdata.numregs = uint16(next)
}

type TempSlice struct {
	regs       []ir.RegKey
	use_counts map[ir.RegKey]int
}

func (p TempSlice) Len() int { return len(p.regs) }
func (p TempSlice) Less(i, j int) bool {
	k1, k2 := p.regs[i], p.regs[j]
	u1 := p.use_counts[k1]
	u2 := p.use_counts[k2]
	return u1 > u2 || (u1 == u2 && k1.Less(k2))
}
func (p TempSlice) Swap(i, j int) { p.regs[i], p.regs[j] = p.regs[j], p.regs[i] }

func (p TempSlice) Sort() []ir.RegKey {
	sort.Sort(p)
	return p.regs
}

// Sort registers by number of uses so that more frequently used registers will
// end up in slots 0-3 or 4-255 and benefit from the shorter instruction forms
// For simplicity, parameter registers are still preserved as is with one exception
func SortAllocateRegisters(irdata *IRWriter) {
	NONE := ir.RegKey{}
	instrs := irdata.Instructions

	use_counts := make(map[ir.RegKey]int)
	for _, instr := range instrs {
		if ins, ok := instr.(*ir.RegAccess); ok {
			use_counts[ins.RegKey] += 1
		}
	}

	regs := append([]ir.RegKey(nil), irdata.initial_args...)

	keys := []ir.RegKey{}
	for k, _ := range use_counts {
		keys = append(keys, k)
	}
	for _, key := range (TempSlice{keys, use_counts}.Sort()) {
		// If key is a param, it was already added at the beginning
		isparam := false
		for _, param := range irdata.initial_args {
			if param == key {
				isparam = true
				break
			}
		}

		if !isparam {
			regs = append(regs, key)
			if key.T.Wide() {
				regs = append(regs, NONE)
			}
		}
	}

	// Sometimes the non-param regsisters are used more times than the param registers
	// and it is beneificial to swap them (which requires inserting code at the
	// beginning of the method to move the value if the param is not unused)
	// This is very complicated to do in general, so the following code only does
	// this in one specific circumstance which should nevertheless be sufficient
	// to capture the majority of the benefit
	// Specificially, it only swaps at most one register, and only in the case that
	// it is nonwide and there is a nonwide parameter in the first 4 slots that
	// it can be swapped with. Also, it doesn't bother to check if param is unused.
	candidate_i := len(irdata.initial_args)
	if candidate_i < 4 {
		candidate_i = 4
	}
	// make sure candidate is valid, nonwide register
	if len(regs) > candidate_i && regs[candidate_i] != NONE {
		candidate := regs[candidate_i]
		if !candidate.T.Wide() && use_counts[candidate] >= 3 {

			for i := 0; i < 4 && i < len(irdata.initial_args); i++ {
				// make sure target is not wide
				if regs[i] == NONE || regs[i+1] == NONE {
					continue
				}

				target := regs[i]
				if use_counts[candidate] > use_counts[target]+3 {
					// swap register assignments
					regs[i], regs[candidate_i] = candidate, target
					// add move instructions at beginning of method
					load := ir.RawRegAccess(i, target.T, false)
					store := ir.NewRegAccess(target.Reg, target.T, true)
					instrs = append([]ir.Instruction{load, store}, instrs...)
					irdata.Instructions = instrs
					break
				}
			}
		}
	}

	// Now generate bytecode from the selected register allocations
	irdata.numregs = uint16(len(regs))
	regmap := map[ir.RegKey]int{}
	for i, v := range regs {
		if v != NONE {
			regmap[v] = i
		}
	}

	for _, instr := range instrs {
		if ins, ok := instr.(*ir.RegAccess); ok && !instr.HasBytecode() {
			ins.CalcBytecode(regmap[ins.RegKey])
		}
	}
}

File: enjarify-rs/src/jvm/optimization/registers.rs
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
use std::cell::{RefCell, RefMut};
use std::cmp::{max, min};
use std::collections::{HashMap, HashSet};
use std::collections::vec_deque::VecDeque;
use std::ops::Deref;
use std::ptr::null;
use std::rc::Rc;


use jvm::ir;
use jvm::writeir::IRWriter;

// Copy propagation - when one register is moved to another, keep track and replace
// all loads with loads from the original register (as long as it hasn't since been
// overwritten). Note that stores won't be removed, since they may still be needed
// in some cases, but if they are unused, they'll be removed in a subsequent pass
// As usual, assume no iincs

// A set of registers that currently are copies of each other.
#[derive(Clone)]
struct CopySet{
    root: ir::RegKey,
    set: HashSet<ir::RegKey>,
    q: VecDeque<ir::RegKey>, // keep track of insertion order in case root is overwritten
}
impl CopySet {
    fn new(key: ir::RegKey) -> CopySet {
        let mut t = CopySet{root: key, set: HashSet::new(), q: VecDeque::new()};
        t.set.insert(key);
        t
    }

    fn add(&mut self, key: ir::RegKey) {
        assert!(self.set.len() > 0);
        self.set.insert(key);
        self.q.push_back(key);
    }

    fn remove(&mut self, key: ir::RegKey) {
        self.set.remove(&key);
        // Heuristic - use oldest element still in set as new root
        while self.q.len() > 0 && !self.set.contains(&self.root) {
            self.root = self.q.pop_front().unwrap();
        }
    }
}

fn ptr(p: Option<&Rc<RefCell<CopySet>>>) -> *const CopySet {
    match p {
        Some(p) => p.deref().borrow().deref() as *const _,
        None => null(),
    }
}

// Map registers to CopySets
#[derive(Default)]
struct CopySetsMap(HashMap<ir::RegKey, Rc<RefCell<CopySet>>>);
impl CopySetsMap {
    fn get(&mut self, key: ir::RegKey) -> RefMut<CopySet> {
        self.0.entry(key).or_insert_with(|| Rc::new(RefCell::new(CopySet::new(key)))).borrow_mut()
    }

    fn load(&mut self, key: ir::RegKey) -> ir::RegKey { self.get(key).root }

    fn clobber(&mut self, key: ir::RegKey) {
        if let Some(v) = self.0.remove(&key) {
            v.borrow_mut().remove(key);
        }
    }

    fn move_(&mut self, src: ir::RegKey, dest: ir::RegKey) -> bool {
        // return false if the corresponding instructions should be removed
        let s_set = ptr(self.0.get(&src));
        let d_set = ptr(self.0.get(&dest));
        if !s_set.is_null() && s_set == d_set {
            // src and dest are copies of same value, so we can remove
            return false;
        }
        if !d_set.is_null() { self.get(dest).remove(dest); }
        self.get(src).add(dest);

        // copy src set over to dest
        let r = self.0.get(&src).unwrap().clone();
        self.0.insert(dest, r);
        true
    }

    fn clone(&self) -> CopySetsMap {
        // overriden to perform deep copy while maintaining aliases
        let mut copies = HashMap::new();
        let mut refs = HashMap::new();
        for (k, v) in &self.0 {
            let c = copies.entry(ptr(Some(&v))).or_insert_with(||
                Rc::new(RefCell::new(v.deref().borrow().clone()))
            );
            refs.insert(*k, c.clone());
        }
        CopySetsMap(refs)
    }
}


pub fn copy_propagation(irdata: &mut IRWriter) {
    let mut replace = HashMap::new();
    let mut single_pred_infos = HashMap::new();

    let mut current = CopySetsMap::default();
    let mut prev: (_,_,Option<ir::RAImpl>) = (0, true, None);

    for (i, instr) in irdata.instructions.iter().enumerate() {
        // reset all info when control flow is merged
        if irdata.is_target(instr.lbl()) {
            let lbl = instr.lbl().unwrap();
            // try to use info if this was a single predecessor forward jump
            let pred_counts = irdata.target_pred_counts[&lbl];
            if !prev.1 && pred_counts == 1 {
                current = single_pred_infos.remove(&lbl).unwrap_or(CopySetsMap::default());
            } else {
                current = CopySetsMap::default();
            }
        } else if let ir::RegAccess(ref data) = instr.sub {
            let key = data.key;
            if data.store {
                // check if previous instr was a load
                if let Some(data2) = prev.2.take() {
                    if !data2.store {
                        if !current.move_(data2.key, key) {
                            replace.insert(prev.0, Vec::new());
                            replace.insert(i, Vec::new());
                        }
                    } else { current.clobber(key); }
                } else {
                    current.clobber(key);
                }
            } else {
                let root_key = current.load(key);
                if key != root_key {
                    assert!(!replace.contains_key(&i));
                    // replace with load from root register instead
                    replace.insert(i, vec![ir::reg_access(root_key.0, root_key.1, false)]);
                }
            }
        } else {
            for target in instr.targets() {
                let lbl = ir::LabelId::DPos(target);
                if irdata.target_pred_counts[&lbl] == 1 {
                    single_pred_infos.insert(lbl, current.clone());
                }
            }
        }

        prev = (i, instr.fallsthrough(),
            if let ir::RegAccess(ref data) = instr.sub {Some(data.clone())} else {None}
        );
    }

    irdata.replace_instrs(replace);
}


pub fn remove_unused_registers(irdata: &mut IRWriter) {
    // Remove stores to registers that are not read from anywhere in the method
    let mut used = HashSet::new();
    for instr in irdata.instructions.iter() {
        if let ir::RegAccess(ref data) = instr.sub {
            if !data.store {
                used.insert(data.key);
            }
        }
    }

    let mut replace = HashMap::new();
    let mut prev_was_replaceable = false;
    for (i, instr) in irdata.instructions.iter().enumerate() {
        if let ir::RegAccess(ref data) = instr.sub {
            if !used.contains(&data.key) {
                assert!(data.store);

                // if prev instruction is load or const, just remove it and the store
                // otherwise, replace the store with a pop
                if prev_was_replaceable {
                    replace.insert(i-1, Vec::new());
                    replace.insert(i, Vec::new());
                } else {
                    let wide = data.key.1.is_wide();
                    replace.insert(i, vec![if wide {ir::pop2()} else {ir::pop()}]);
                }
            }

            prev_was_replaceable = !data.store; // loads are ok
        } else { prev_was_replaceable = instr.is_constant(); }
    }
    irdata.replace_instrs(replace);
}

// Allocate registers to JVM registers on a first come, first serve basis
// For simplicity, parameter registers are preserved as is
pub fn simple_allocate_registers(irdata: &mut IRWriter) {
    let mut regmap: HashMap<_,_> = irdata.initial_args.iter().enumerate().map(|(i, v)| (v, i as u16)).collect();
    let mut nextreg = irdata.initial_args.len() as u16;

    for instr in &mut irdata.instructions {
        if let ir::RegAccess(ref data) = instr.sub {
            let reg = *regmap.entry(&data.key).or_insert_with(|| {
                let t = nextreg;
                nextreg += 1;
                if data.key.1.is_wide() { nextreg += 1; }
                t
            });
            instr.bytecode = Some(data.calc_bytecode(reg));
        }
    }
    irdata.numregs = Some(nextreg);
}

// Sort registers by number of uses so that more frequently used registers will
// end up in slots 0-3 or 4-255 and benefit from the shorter instruction forms
// For simplicity, parameter registers are still preserved as is with one exception
pub fn sort_allocate_registers(irdata: &mut IRWriter) {
    let mut use_counts = HashMap::new();
    {
        for instr in &mut irdata.instructions {
            if let ir::RegAccess(ref data) = instr.sub {
                *use_counts.entry(data.key).or_insert(0) += 1;
            }
        }
    }

    let mut regs = irdata.initial_args.clone();
    let mut rest: Vec<_> = use_counts.keys().map(|&k| k).collect();
    rest.sort_by_key(|k| (!use_counts[k], *k)); // ! -> bitwise not -> sort decreasing
    for key in rest {
        // If key is a param, it was already added at the beginning
        if !irdata.initial_args.contains(&key) {
            regs.push(key);
            if key.1.is_wide() {
                regs.push(ir::INVALID_KEY);
            }
        }
    }

    // make sure any key we might query is in the dict
    for key in &irdata.initial_args { use_counts.entry(*key).or_insert(0); }

    // Sometimes the non-param regsisters are used more times than the param registers
    // and it is beneificial to swap them (which requires inserting code at the
    // beginning of the method to move the value if the param is not unused)
    // This is very complicated to do in general, so the following code only does
    // this in one specific circumstance which should nevertheless be sufficient
    // to capture the majority of the benefit
    // Specificially, it only swaps at most one register, and only in the case that
    // it is nonwide and there is a nonwide parameter in the first 4 slots that
    // it can be swapped with. Also, it doesn't bother to check if param is unused.
    let nargs = irdata.initial_args.len();
    let candidate_i = max(4, nargs);
    // make sure candidate is valid, nonwide register
    if candidate_i < regs.len() && regs[candidate_i] != ir::INVALID_KEY {
        let candidate = regs[candidate_i];
        if !candidate.1.is_wide() && use_counts[&candidate] >= 3 {
            for i in 0..min(4, nargs) {
                // make sure target is not wide
                if regs[i] == ir::INVALID_KEY || regs[i+1] == ir::INVALID_KEY { continue; }

                let target = regs[i];
                if use_counts[&candidate] > use_counts[&target] + 3 {
                    // swap register assignments
                    regs[i] = candidate;
                    regs[candidate_i] = target;
                    // add move instructions at beginning of method
                    let load = ir::raw_access(i as u16, target.1, false);
                    let store = ir::reg_access(target.0, target.1, true);
                    // todo - can these be inserted simulatenously to avoid extra shift?
                    irdata.instructions.insert(0, store);
                    irdata.instructions.insert(0, load);
                    // println!("moving arg {} -> {}", i, target.0);
                    break;
                }
            }
        }
    }
    // println!("regs {:?}", regs);

    // Now generate bytecode from the selected register allocations
    let regmap: HashMap<_,_> = regs.iter().enumerate().map(|(i, v)| (v, i as u16)).collect();
    for instr in &mut irdata.instructions {
        if instr.bytecode.is_none() {
            if let ir::RegAccess(ref data) = instr.sub {
                instr.bytecode = Some(data.calc_bytecode(regmap[&data.key]));
            }
        }
    }
    irdata.numregs = Some(regs.len() as u16);
}
